### PR TITLE
Fix Hebrew highlighting regex

### LINKED ISSUE #3476: [BUG] Erroneous syntax highlighting in Hebrew. CHARACTER regex without Hebrew characters
**Describe the bug**
In Hebrew, there are two separate keywords: "ב" (at) and "בתוך" (in). When typing the keyword "בתוך", the editor recognizes "ב" as a complete word, and doesn't highlight the rest of the word.
I believe this is down to the fact that the constant `CHARACTER` in the file [highlighting/definition.py](https://github.com/Felienne/hedy/blob/main/highlighting/definition.py), doesn't include Hebrew characters. It should include the range א-ת (Aleph א through Tav ת).

**Add a screenshot (optional)**
Observed in level 10
image

**Expected behavior**
The entire word "בתוך" should be highlighted as a keyword, not just the first letter "ב".

### PR DESCRIPTION
**Pull request by o11k**
Fixes #3476

The regex `CHARACTER` from [highlighting/definition.py](https://github.com/Felienne/hedy/blob/main/highlighting/definition.py) now includes Hebrew characters. The editor no longer recognizes Hebrew characters as end-of-word.

Before (level 11):
image

After:
image

### CODE REVIEW

**Comment by thjazi39** [comment]
I validate the modification that has been made to support the coloring of Hebrew characters. Maybe add some tests ?

**Comment by o11k** [comment]
@Felienne You're right. `re` doesn't support Unicode categories, but the external `regex` does.
However, the pattern `[^\W\d_]` *almost* means "letter". More specifically, it matches characters from the following categories:
- Uppercase letter (Lu)
- Lowercase letter (Ll)
- Titlecase letter (Lt)
- Modifier letter (Lm)
- Other letter (Lo)
- Letter number (Nl)
- Other number (No)

Sadly, there are 686 characters from these categories, that this pattern doesn't match. All of them are very obscure, and most of them aren't even printable using Hedy's font.
https://gist.github.com/o11k/c646b5a09e3acee62f65e9c0c8cab773

There are basically 3 paths forward now:
1. Keep adding character ranges as needed - easy but a constant WIP
2. `[^\W\d_]` - my favorite :)
3. Use an external module like `regex` - also good.

**Comment by Felienne** [comment]
Thanks for looking into this @o11k!!

> There are basically 3 paths forward now:
> 
> 1. Keep adding character ranges as needed - easy but a constant WIP
> 2. `[^\W\d_]` - my favorite :)
> 3. Use an external module like `regex` - also good.

I would prefer not 1 because that keeps showing buggy-looking code to users. I am totally ok with 2, what is the reason it is your favorite and do you see any downsides? If it matches a bit more like you say, that does not seem to be a problem to me, any other reasons we might not want that?

**Review by Felienne — CHANGES_REQUESTED** [review]
Ideally we fix this at once for all other non-Latin languages too!

**Comment by o11k** [comment]
@Felienne How embarrassing... I missed some important details:

- `CHARACTER` includes numbers and an underscore, so we could have just replaced it with a `\w`, if not for the next fact.
- This regular expression is passed to the Ace editor, in Javascript (not Python). AFAIK, the only way to make Javascript match non-English letters and numbers, is to use the [Unicode flag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicode), which the Ace editor completely filters out. I tried it.

Unless we want to hack/fork/replace Ace, our best choice might be sticking to explicit character ranges...

**Comment by Felienne** [comment]
> Unless we want to hack/fork/replace Ace, our best choice might be sticking to explicit character ranges...

Ah yes, now that you mention it, I know there was a reason we did not just do `\w`. Shall we then explicitly add "firstletter-lastletter" for our non-Latin langs, so that is (I think?)

Currently, I think: Thai, Bengali, Chinese, Farsi, Punjabi, Turkish, Russian, Ukranian, Greek (but maybe I forgot one, I used to be able to name all, but I lost count at one point haha!)

This might also requires diving into languages, f.e. do Russian and Ukranian share an alfabet? Turkish has a lot of Latin letters, but what extra's do they need etc...

I 100% understand if you do not want to fix this further @o11k! We can also pick it up if this is too much work for you, let us know (here or in Discord).

**Comment by jpelay** [comment]
> * ludes numbers and an underscore, so we could have just replaced it wit



> @Felienne How embarrassing... I missed some important details:
> 
>     * `CHARACTER` includes numbers and an underscore, so we could have just replaced it with a `\w`, if not for the next fact.
> 
>     * This regular expression is passed to the Ace editor, in Javascript (not Python). AFAIK, the only way to make Javascript match non-English letters and numbers, is to use the [Unicode flag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicode), which the Ace editor completely filters out. I tried it.
> 
> 
> Unless we want to hack/fork/replace Ace, our best choice might be sticking to explicit character ranges...

Yes, we're in the same place right now. I was just trying this, and the ACE editor filters the u flag!

**Comment by jpelay** [comment]
@Felienne I solved the problem by doing a few things:

1. Changing the `CHARACTER` regex to the one used in the grammars.
2. Changing the regex library used in the highlighter module from`re` to `regex`, because `regex` does accept unicode categories.
3. Updating ace.js to the last version.
4. Adding the `unicode: true` attribute to all the rules in the highlighter, this will tell ace that it needs to add the unicode flag to the regexes.

The code in the issue now looks like this:
image

And the one in level 11 looks like this:
image

**Comment by o11k** [comment]
It works! This is probably the best solution. Thank you @jpelay !

**Review by Felienne — APPROVED** [review]
Just checked it and yes it works for Hebrew and also for other langs like Chinese: 

image

File: requirements.txt [hunk 1]
```diff
@@ -24,3 +24,4 @@ parameterized==0.8.1
 Flask-Babel==2.0.0
 iso3166~=2.0.2
 turtlethread>=0.0.6
+pre-commit==2.20.0
```
[the remark this answers is not in the document]
Comment (jpelay) [reply requirements.txt]: No it isn't intentional!

[the remark this answers is not in the document]
Comment (jpelay) [reply requirements.txt]: Hi! No it wasn't, but notice that I reverted that change and won't be committed!

**Pull request merged**